relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/server/CogRPCService.kt

```diff
@@ -0,0 +1,97 @@
+package tech.relaycorp.relaynet.cogrpc.server
+
+import com.google.protobuf.ByteString
+import io.grpc.stub.StreamObserver
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import tech.relaycorp.relaynet.CargoRelay
+import tech.relaycorp.relaynet.CargoRelayServer
+import tech.relaycorp.relaynet.cogrpc.CargoDelivery
+import tech.relaycorp.relaynet.cogrpc.CargoDeliveryAck
+import tech.relaycorp.relaynet.cogrpc.CargoRelayGrpc
+import java.util.logging.Level
+import java.util.logging.Logger
+
+class CogRPCService(
+    private val coroutineScope: CoroutineScope,
+    private val connectionService: CargoRelayServer.ConnectionService
+) : CargoRelayGrpc.CargoRelayImplBase() {
+
+    override fun deliverCargo(responseObserver: StreamObserver<CargoDeliveryAck>) =
+        object : StreamObserver<CargoDelivery> {
+            override fun onNext(cargoDelivery: CargoDelivery) {
+                coroutineScope.launch {
+                    logger.info("deliverCargo next")
+                    connectionService.deliverCargo(cargoDelivery.toMessageReceived())
+                    logger.info("deliverCargo next ack")
+                    responseObserver.onNext(cargoDelivery.toAck())
+                }
+            }
+
+            override fun onError(throwable: Throwable) {
+                logger.log(Level.WARNING, "deliverCargo error", throwable)
+            }
+
+            override fun onCompleted() {
+                logger.info("deliverCargo complete")
+                coroutineScope.launch {
+                    responseObserver.onCompleted()
+                }
+            }
+        }
+
+    override fun collectCargo(responseObserver: StreamObserver<CargoDelivery>): StreamObserver<CargoDeliveryAck> {
```

Comment:
Note that my comments on log levels in the `deliverCargo` method also apply here.